### DISCUSSION TITLE
Reverse URL lookups between APIRotuers

### DISCUSSION THREAD

**Discussion by Fingel** [fastapi/fastapi#7086] [Questions] [answered]
* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Environment

* OS: macOS
* FastAPI 0.62.0
* Python 3.9

### Description

Reverse url lookups are something that doesn't seem to be documented very well for FastAPI, but they are [supported by Starlette](https://www.starlette.io/routing/#reverse-url-lookups) and extremely helpful to avoid hardcoding url paths. Hardcoding paths can become a real nuisance, especially in the starting phases of a project (and most projects starting with FastAPI are probably new!) as you shuffle around your routes in order to match the desired API design. Not only that, but they are almost necessary if you are writing tests!

I did find this issue: https://github.com/tiangolo/fastapi/issues/199 which goes over using `app.url_path_for` which is helpful for simple apps, but if you are using multiple routers it does not seem to be sufficient.

Ideally, we could use  APIRouter.url_path_for('other:some_route') to do reverse url lookups, as that does seem to be supported in Starlette, but it fails in FastAPI.

Below is a simple test case with the failing test that I believe should pass.

```python
from fastapi import FastAPI, APIRouter
from fastapi.testclient import TestClient

app = FastAPI()

foo_router = APIRouter(prefix='/foo')


@foo_router.get('/sub')
async def foo_sub():
    return {'message': 'Hello from foo sub'}


bar_router = APIRouter(prefix='/bar')


@bar_router.get('/sub')
async def bar_sub():
    return {'message': 'Hello from bar sub'}

app.include_router(foo_router)
app.include_router(bar_router)

client = TestClient(app)


def test_url_lookup_same_router():
    url = foo_router.url_path_for("foo_sub")
    response = client.get(url)
    assert response.json() == {"message": "Hello from foo sub"}


def test_url_lookup_other_router():
    url = bar_router.url_path_for("foo:foo_sub")
    response = client.get(url)
    assert response.json() == {"message": "Hello from foo sub"}
```

The last test fails with `starlette.routing.NoMatchFound`

**Comment by Mause** [comment]
What are you referring to for that colon syntax in the url_path_for call?

In any case, this is probably the closest to what you want:


```python

@bar_router.get('/sub')
async def bar_sub(request: Request):
    return {'message': request.url_for('foo_sub')}
```

**Comment by Fingel** [comment]
The colon syntax in from the Starlette docs, it looks like routes can be namespaced, which is handy:  

https://www.starlette.io/routing/#reverse-url-lookups

I didn't know about Request.url_for(), that's very handy. Thank you! Any thoughts on how we might be able to reverse URLs from within tests where we don't have access to a request yet, for example, to make a GET request on a specific endpoint?

**Comment by Mause** [comment]
Ah, that isn't very well documented :/

You don't really need to use Request.url_for unless you want a URL to be relative to the current route - app.url_for_path exists as well.

Perhaps if you explained why you are trying to generate URLs outside of a request scope in a test we could better help?

**Comment by Fingel** [comment]
Hi,

I've found `app.url_for_path` is sufficient for testing. It would be nice to have some kind of namespacing, so you could lookup "foo:list" and "bar:list",  because as of now you have to ensure that every route function has a unique name. That's fine, just something to be careful of.

**Comment by hmoravec** [comment, marked as answer]
It is possible to specify route name via attribute `name` in routes as mentioned in https://github.com/tiangolo/fastapi/issues/3119. If not specified, function name is used. The route name is then used in `url_for` and `url_for_path` methods. Unfortunately it is also shown as endpoint name in Swagger UI but if one chooses reasonable naming pattern, it is possible to define custom route names usable both for reverse URL lookups and as endpoint names in Swagger UI.

**Comment by tiangolo** [comment]
Thanks for the help here @Mause and @hmoravec  ! 👏  🙇 

Thanks for reporting back and closing the issue @Fingel  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
